Add takewhile, dropwhile to iterators

o314 opened the pull request:
Constating that python, ruby, rust, elixir all offer takewhile, dropwhile

| language | take while| drop while |
|---|---|---|
| python | [takewhile](https://docs.python.org/3.7/library/itertools.html#itertools.takewhile) | [dropwhile](https://docs.python.org/3.7/library/itertools.html#itertools.dropwhile) |
| ruby | [take_while](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-take_while) | [drop_while](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-drop_while) |
| rust | [take_while](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take_while) | [skip_while](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip_while) |
| elixir | [take_while](https://hexdocs.pm/elixir/Enum.html#take_while/2) | [drop_while](https://hexdocs.pm/elixir/Enum.html#drop_while/2) |

And also dotnet, java, scala !

I think Julia should have them too.

IterTools.jl, and rust have been consulted.

o314 committed:
Add takewhile, dropwhile to iterators

o314 committed:
Add takewhile, dropwhile tests to iterators tests

ararslan commented on base/iterators.jl:
```diff
@@ -649,6 +649,52 @@ end
+    pred
+end
+
+iterate(ibl::TakeWhile{I},itr...) where {I} = begin
```
```suggestion
function iterate(ibl::TakeWhile{I},itr...) where {I}
```

ararslan commented on base/iterators.jl:
```diff
@@ -649,6 +649,52 @@ end
+    DropWhile(xs::I,pred) where {I} = new{I}(xs,pred,false)
+end
+
+iterate(ibl::DropWhile{I},itr...) where {I} = begin
```
```suggestion
function iterate(ibl::DropWhile{I},itr...) where {I}
```

ararslan commented on test/iterators.jl:
```diff
@@ -149,6 +149,30 @@ for xs in Any["abc", [1, 2, 3]]
+# takewhile
+# --------
+@testset "takewhile" begin
+    @test takewhile(x -> x<4,1:10) |> collect == [1,2,3]
```
```suggestion
    @test collect(takewhile(x -> x<4,1:10)) == [1,2,3]
```
And the rest of these in this file. Won't comment for each.

ararslan replied:
The others haven't been addressed

o314 replied:
received. i have some work ahead for now, i will fix it tomorrow.

JeffBezanson commented on base/iterators.jl:
```diff
@@ -649,6 +649,52 @@ end
+takewhile(pred::Function,xs) = TakeWhile(xs,pred)
+struct TakeWhile{I}
+    xs::I
+    pred
```
There should be a type parameter for this.

ararslan commented on test/iterators.jl:
```diff
@@ -149,6 +149,30 @@ for xs in Any["abc", [1, 2, 3]]
+    @test takewhile(x -> x<4,1:10) |> collect == [1,2,3]
+    @test takewhile(x -> x<4,Iterators.countfrom(1)) |> collect == [1,2,3]
+    @test takewhile(x -> x<4,5:10) |> collect == []
+    @test takewhile(x -> true,5:10) |> collect == 5:10 |> collect
```
```suggestion
    @test collect(takewhile(x -> true,5:10)) == 5:10
```
No need to materialize ranges for comparison.

JeffBezanson commented on base/iterators.jl:
```diff
@@ -649,6 +649,52 @@ end
+
+# takewhile
+
+takewhile(pred::Function,xs) = TakeWhile(xs,pred)
```
I would prefer defining the struct first; that's the style used by the rest of the file.

JeffBezanson commented on base/iterators.jl:
```diff
@@ -649,6 +649,52 @@ end
+mutable struct DropWhile{I}
+    xs::I
+    pred
+    fltd::Bool
```
This should be part of the iteration state instead of making the struct mutable.

ararslan commented on test/iterators.jl:
```diff
@@ -149,6 +149,30 @@ for xs in Any["abc", [1, 2, 3]]
+    @test takewhile(x -> x<4,5:10) |> collect == []
+    @test takewhile(x -> true,5:10) |> collect == 5:10 |> collect
+    @test takewhile(isodd,[1,1,2,3]) |> collect == [1,1]
+    @test takewhile(<(3),[1,1,2,3]) |> s->takewhile(<(2),s) |> collect == [1,1]
```
```suggestion
    @test collect(takewhile(<(2), takewhile(<(3), [1,1,2,3]))) == [1,1]
```

ararslan commented on test/iterators.jl:
```diff
@@ -149,6 +149,30 @@ for xs in Any["abc", [1, 2, 3]]
+@testset "dropwhile" begin
+    @test dropwhile(x -> x<4, 1:10) |> collect == 4:10 |> collect
+    @test (dropwhile(x -> x<4, 1:10) |> collect) isa Vector{Int}
+    @test dropwhile(x -> x<4, []) |> collect == [] |> collect
```
```suggestion
    @test collect(dropwhile(x -> x<4, [])) == []
```
No need to `collect` an `Array`.

ararslan reviewed (dismissed):
Lots of stylistic fixes needed but I agree that the functionality would be nice to have.

o314 committed:
Apply suggestions from code review

Fix stylistic suggestions from ararslan

Co-Authored-By: Alex Arslan <ararslan@comcast.net>

o314 committed:
Fix takewhile changes request from JeffBezanson

o314 committed:
Fix dropwhile changes request from JeffBezanson

JeffBezanson commented on base/iterators.jl:
```diff
@@ -649,6 +649,51 @@ end
+
+takewhile(pred,xs) = TakeWhile(xs,pred)
+
+function iterate(ibl::TakeWhile{I,P},itr...) where {I,P}
```
```suggestion
function iterate(ibl::TakeWhile, itr...)
```

JeffBezanson commented on base/iterators.jl:
```diff
@@ -649,6 +649,51 @@ end
+    y
+end
+
+IteratorSize(ibl::TakeWhile{I,P}) where {I,P} = SizeUnknown()
```
```suggestion
IteratorSize(ibl::Type{<:TakeWhile}) = SizeUnknown()
```

JeffBezanson commented on base/iterators.jl:
```diff
@@ -649,6 +649,51 @@ end
+    y !== nothing ? (y[1],(true,y[2])) : nothing
+end
+
+IteratorSize(ibl::DropWhile{I,P}) where {I,P} = SizeUnknown()
```
```suggestion
IteratorSize(::Type{<:DropWhile}) = SizeUnknown()
```

o314 replied:
How do i miss ::Type{}, sorry

ararslan commented:
Should rebase on master to get the whitespace fix. Also `takewhile` and `dropwhile` should be added to the Iterators submodule exports along with the other functions there. They need docstrings with a compat annotation and should be added to the manual with the other iterators.

JeffBezanson commented on base/iterators.jl:
```diff
@@ -649,6 +649,51 @@ end
+        iterate(ibl.xs, itr...) :
+        iterate(Iterators.filter(!ibl.pred, ibl.xs),itr...)
+
+    y !== nothing ? (y[1],(true,y[2])) : nothing
```
Hmm, I was wrong before --- this doesn't even need to be part of the state, and can be handled simply by doing a while loop at the beginning of iteration.

o314 replied:
all good, i had a test failure and it is a race that happens when two dropwhile calls were nested. this should help to fix. i have some work ahead for now, i will fix it tomorrow.

o314 committed:
Apply suggestions from code review

Relax unionall annotations

Co-Authored-By: Jeff Bezanson <jeff.bezanson@gmail.com>

ararslan commented on base/iterators.jl:
```diff
@@ -649,6 +649,51 @@ end
+
+# takewhile
+
+struct TakeWhile{I,P<:Function}
```
The predicate doesn't necessarily need to be a function, for example it's not unreasonable to do something like `takewhile(Bool, [1, 1, 0, 1])`.

JeffBezanson replied:
Eh, I would seriously question whether that's a good way to write that... But yes, `Filter` for example does not constrain its function.

o314 replied:
this is the compress function of python itertools. it exists and seems to be used!
the name is misleading however, i prefer filter(mask) with mask a bitvector.
it is useful, to do bitvector ops. a lot of sat stuff go this way nowaday

o314 replied:
One will lands to using a unary closure like with find. may be a callable.

o314 committed:
Fix type annotation, impedance between slat and =nothing

ararslan commented on base/iterators.jl:
```diff
@@ -649,6 +649,51 @@ end
+    pred::P
+end
+
+takewhile(pred,xs) = TakeWhile(xs,pred)
```
The function should probably come first in the type constructor as well; this reversal is kind of weird.

o314 commented:
Thanks for the feedback and openness !
As stated, i have some work ahead for now, i will fix remaining parts asap,
tomorrow the better.

o314 committed:
Fix some takewhile issues

- [x] fix order of iterate? / ctor args / struct fields
- [x] review (leaning of) type annotation

o314 committed:
Fix some takewhile test issues

- [x] replace `x->x<k` by `<(k)` @ test
- [x] replace `x->some_const` by `_->some_const` @ test
- [x] remove excedentary collect @ test

o314 committed:
Fix some dropwhile issue

- [x] fix order of ctor args / struct fields
- [y] review (leaning of) type annotation 
    - [x] redo explicitation @ IteratorSize(::Type{DropWhile}).  
    otherwise MethodError: no method matching length was raised

o314 committed:
Fix some dropwhile test issues

- [x] replace `x->x<k` by `<(k)` @ test
- [x] replace `x->some_const` by `_->some_const` @ test
- [x] remove excedentary collect @ test
- [/] remove pipeline @ test

o314 committed:
Fix dropwhile algo issues

Choice has been made to reimplement filter with a recoded while pred loop.
State partition is determined only by the emptiness of iterator

One could use Iterators.filter with Iterators.Stateful but filter erases last elem state and stateful did not revert propertly due to nextvalstate/taken machinery. Stateful needs some cleanup.

o314 committed:
Fix dropwhile algo test issues

KristofferC commented on base/iterators.jl:
```diff
@@ -649,6 +649,58 @@ end
+    end
+end
+
+IteratorSize(::Type{DropWhile{I,P}}) where {I,P} = SizeUnknown()
```
Could write these and the others as

```suggestion
IteratorSize(::Type{<:DropWhile}) = SizeUnknown()
```

to not have to mention I and P everywhere.

o314 replied:
Jeff indicates that yesterday, i first fix it. But that raised a MethodError: no method matching length
May be because my code was not in base?
I suggest to let integration run and retry a fix.

o314 commented:
Possible Enhancement

- [ ] optimize nested `TakeWhile(TakeWhile())`, `DropWhile(DropWhile())`
    `DropWhile(TakeWhile())`, `TakeWhile(DropWhile())`

halleysfifthinc commented on base/iterators.jl:
```diff
@@ -649,6 +649,58 @@ end
+end
+
+IteratorSize(::Type{DropWhile{I,P}}) where {I,P} = SizeUnknown()
+eltype(::Type{DropWhile{I,P}}) where {I,P} = eltype(I)
```
This line and others would be better as
```suggestion
eltype(::Type{DropWhile{I}}) where {I} = eltype(I)
```
since `P` isn't used

KristofferC replied:
That doesn't work.

o314 committed:
Microfix of export takewhile, dropwhile

JeffBezanson commented on base/iterators.jl:
```diff
@@ -649,6 +649,58 @@ end
+dropwhile(pred,itr) = DropWhile(pred,itr)
+
+function iterate(ibl::DropWhile,itr...)
+    if isempty(itr)
```
I think it would be more elegant to write this as two methods instead of branching on the number of arguments.

tkf commented on base/iterators.jl:
```diff
@@ -649,6 +649,58 @@ end
+
+struct TakeWhile{I,P<:Function}
+    pred::P
+    xs::I
```
Could you be consistent about naming of the inner iterator?

```julia
julia> lines = readlines(`sh -c "grep --line-number -A5 -E '^struct' iterators.jl | grep -E -e struct -e '::[A-Z]\$'"`)
32-element Array{String,1}:
 "77:struct Reverse{T}"
 "78-    itr::T"
 "106:struct Enumerate{I}"
 "107-    itr::I"
 "166:struct Pairs{K, V, I, A} <: AbstractDict{K, V}"
 "167-    data::A"
 "168-    itr::I"
 "263:struct Zip{Is<:Tuple}"
 "393:struct Filter{F,I}"
 "394-    flt::F"
 "395-    itr::I"
 "445:struct Rest{I,S}"
 "446-    itr::I"
 "447-    st::S"
 "504:struct Count{S<:Number}"
 "505-    start::S"
 "506-    step::S"
 "537:struct Take{I}"
 "538-    xs::I"
 "594:struct Drop{I}"
 "595-    xs::I"
 "653:struct Cycle{I}"
 "654-    xs::I"
 "691:struct Repeated{O}"
 "692-    x::O"
 "726:struct ProductIterator{T<:Tuple}"
 "727-    iterators::T"
 "848:struct Flatten{I}"
 "849-    it::I"
 "932:struct PartitionIterator{T}"
 "933-    c::T"
 "957:struct IterationCutShort; end"

julia> count(x -> occursin("xs::", x), lines)
3

julia> count(x -> occursin("itr::", x), lines)
5
```

`xs` is popular but using `itr` would be most consistent with everything else.

(I know this is an implementation detail but it helps me implementing https://tkf.github.io/Transducers.jl/dev/manual/#Transducers.eduction)

o314 replied:
The problem is that there is a lot of inconsistency between iterator as `iterable` and `iterator` as `state` (iteration state) eg arg1 and arg2 of `iterate/2`. In `C#` there is clearly `IEnumerable` and `IEnumerator` interface. I am not sure `Linq` would be so great otherwise.

Most people are only data poweruser and are not so aware. Julia should inforce this too i think. Following this advise will lead to augment confusion, i don't think it's a good way.

I wrote my code with either `ibl`/`itr` either `xs`/`itr` when `ibl` is indexable. It was relaxed for Base. But it works ... everywhere, so far :smiley:

tkf replied:
> The problem is that there is a lot of inconsistency between iterator as `iterable` and `iterator` as `state` (iteration state)

I know the difference. But please observe that all the `struct`s above with `xs` and `itr` are the same thing; i.e., they are fed to the first argument of `iterate`. Consider `itr` an abbreviation of iterable if that's your preference.

> In `C#` there is clearly `IEnumerable` and `IEnumerator` interface.

I'm not familiar with C# but this seems to be irrelevant in Julia because the iterator interface in Julia is formalized in terms of methods without side-effect; i.e., there is no mutable "iterator" object. (My understanding is that IEnumerable and IEnumerator in C# are iterable and iterator in Python)

o314 replied:
I don't understand your point. You ask me to be consistent on a change over 5% of iterators constating base isn't about over 95%. Names used are the same as take/while, the nearer in semantics. So yes this is consistent.

It could be a good thinks to homogenize first. But that's not done and that's not the purpose of adding takewhile, dropwhile. May be submit a pr for that. May be better, participate to an iterator julep https://github.com/JuliaLang/julia/issues/33239

Same pattern in python, java, c# it seems for distinction between `iterable`/`iterator` eg. `enumerable`/`enumerator`.

By the way, sorry to say that bikeshedding for hour is not the way i will develop open source software. Too much work, too few time to enjoy life

tkf replied:
It's not bikeshedding. It affects code dealing with iterators, as I mentioned in the first comment.

JeffBezanson replied:
The field names are already not consistent, so I'd suggest just changing them all in a separate PR after this is merged.

I don't understand how it's useful to access the iterator "inside" another. Does that operation mean anything in terms of the iteration abstraction?

tkf replied:
Yes, renaming them and/or adding a proper API (e.g., `parent`) makes sense.

> Does that operation mean anything in terms of the iteration abstraction?

It's certainly unnecessary for iteration abstraction.  I'm using it for converting "iterator transforms" like `Base.Generator` to transducers which unfortunately requires to get into the concrete implementation details:

https://github.com/tkf/Transducers.jl/blob/3cd6463ea08cd475b2020a6e8deca244a105d20e/src/comprehensions.jl#L89-L102

FYI my another comment on iterator-transducer conversion: https://github.com/JuliaLang/julia/pull/33027#issuecomment-524529959

tkf replied:
Yes, renaming them and/or adding a proper API (e.g., `parent`) makes sense.

> Does that operation mean anything in terms of the iteration abstraction?

It's certainly unnecessary for iteration abstraction.  I'm using it for converting "iterator transforms" like `Base.Generator` to transducers which unfortunately requires to get into the concrete implementation details:

https://github.com/tkf/Transducers.jl/blob/3cd6463ea08cd475b2020a6e8deca244a105d20e/src/comprehensions.jl#L89-L102

FYI my another comment on iterator-transducer conversion: https://github.com/JuliaLang/julia/pull/33027#issuecomment-524529959

o314 committed:
Update dropwhile

- [x] split `iterate` according to presence of `itr` or not
- [x] relax type annotation of `IteratorSize`

JeffBezanson commented:
Code looks good now, thanks. Just needs news and docs.

nalimilan commented on test/iterators.jl:
```diff
@@ -149,6 +149,29 @@ for xs in Any["abc", [1, 2, 3]]
     @test take(drop(take(xs, 3), 1), 1) === take(drop(xs, 1), 1)
 end
 
+# takewhile
```
This looks redundant with the name of the testset. Same below.

o314 replied:
done

o314 committed:
Update docs of takewhile, dropwhile

- [x] set docs of takewhile, dropwhile
    - [x] set compat

o314 committed:
Update test/iterators.jl

- [x] fix takewhile, dropwhile test comment

o314 commented:
# news update

## @ new library functions

- `takewhile()` function returns an iterator that generates element from `iter` as long as `pred` is true (#33437).
- `dropwhile()` function returns an iterator that drops element from `iter` as long as `pred` is true, afterwards, returns every element (#33437).

o314 commented:
# compat update

compat has been set to Julia 1.3

o314 committed:
Fix space bug ! in jldoctest of takewhile, dropwhile

ararslan commented on base/iterators.jl:
```diff
@@ -649,6 +649,105 @@ end
+afterwards, drops every element.
+
+!!! compat "Julia 1.3"
+     This function requires at least Julia 1.3.
```
```suggestion
    This function requires at least Julia 1.3.
```

ararslan commented on base/iterators.jl:
```diff
@@ -649,6 +649,105 @@ end
+afterwards, returns every element.
+
+!!! compat "Julia 1.3"
+     This function requires at least Julia 1.3.
```
```suggestion
    This function requires at least Julia 1.3.
```

ararslan commented:
Julia 1.3 is feature frozen, so the compatibility notice here should be set to 1.4. As for a NEWS item, I'd just say
```
* `takewhile` and `dropwhile` have been added to the Iterators submodule ([#33437]).
```

o314 committed:
Update base/iterators.jl

Co-Authored-By: Alex Arslan <ararslan@comcast.net>

o314 committed:
Update base/iterators.jl

Co-Authored-By: Alex Arslan <ararslan@comcast.net>

o314 committed:
Update iterators.jl

o314 commented:
I've updated the news and update sections according to your remarks.

ararslan commented:
Looks like the NEWS file hasn't been updated.

o314 committed:
Update NEWS.md

ararslan commented on NEWS.md:
```diff
@@ -29,6 +29,7 @@ New library functions
 * The `tempname` function now takes a `cleanup::Bool` keyword argument defaulting to `true`, which causes the process to try to ensure that any file or directory at the path returned by `tempname` is deleted upon process exit ([#33090]).
 * The `readdir` function now takes a `join::Bool` keyword argument defaulting to `false`, which when set causes `readdir` to join its directory argument with each listed name ([#33113]).
 * The new `only(x)` function returns the one-and-only element of a collection `x`, and throws an `ArgumentError` if `x` contains zero or multiple elements. ([#33129])
+* takewhile and dropwhile have been added to the Iterators submodule ([#33437]).
```
```suggestion
* `takewhile` and `dropwhile` have been added to the Iterators submodule ([#33437]).
```

o314 committed:
Update NEWS.md

Co-Authored-By: Alex Arslan <ararslan@comcast.net>

o314 commented:
A `filter(bitvector,itr)` has been evocated (think [compress @ python](https://docs.python.org/3/library/itertools.html#itertools.compress)).

A oneliner can do the thing.
(May be two oneliners)

```julia
# zip variant. clone of python compress
@inline Base.Iterators.filter(markers::Union{Vector{Bool},BitVector}, s) =
    (i for (m,i) in zip(markers,s) if m)

# enumerate variant. julia thing
@inline Base.Iterators.filter(markers::Union{Vector{Bool},BitVector}, s) =
    (i for (n,i) in enumerate(s) if markers[n])
```

## Benches

Bench initialization
```julia
using Random
bs = bitrand(10_000);
is = rand(1:100,10_000)
```

### Bench of filter (zip variant)

```
julia> @benchmark Iterators.filter(bs, is)
BenchmarkTools.Trial:
  memory estimate:  80 bytes
  allocs estimate:  4
  --------------
  minimum time:     35.092 ns (0.00% GC)
  median time:      39.066 ns (0.00% GC)
  mean time:        54.293 ns (23.81% GC)
  maximum time:     41.544 μs (99.83% GC)
  --------------
  samples:          10000
  evals/sample:     1000
```

### Bench of filter (enumerate variant)
    
```
julia> @benchmark Iterators.filter(bs, is)
BenchmarkTools.Trial:
  memory estimate:  80 bytes
  allocs estimate:  4
  --------------
  minimum time:     31.782 ns (0.00% GC)
  median time:      35.423 ns (0.00% GC)
  mean time:        50.731 ns (25.69% GC)
  maximum time:     42.052 μs (99.82% GC)
  --------------
  samples:          10000
  evals/sample:     1000
```

The enumerate variant seems 10-15% faster.  
I don't know if we should have something in `Base` for this.
